# Getting our data ready for ML via sklearn

Yeah, I know, it feels kind of difficult... <br>
![](https://media.giphy.com/media/1eBQ8eo1AYyGc/giphy.gif) ![](https://media.giphy.com/media/hSUAf0zd1u56U/giphy.gif)

See? I don't always use memes :)
Best wishes to you!

## Standard imports

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

### Our objectives 
* Split data into features (x usually) and labels (y usually)
* Filling (also called "Imputing") or disregarding missing values
* Converting non-numerical values to numerical values (a.k.a Feature encoding)

In [18]:
# Let's begin with the heart disease dataset
hd = pd.read_csv("https://raw.githubusercontent.com/ineelhere/Machine-Learning-and-Data-Science/master/scikit-learn/heart-disease.csv")
hd.head(3)

age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  slope  \
0   63    1   3       145   233    1        0      150      0      2.3      0   
1   37    1   2       130   250    0        1      187      0      3.5      0   
2   41    0   1       130   204    0        0      172      0      1.4      2   

   ca  thal  target  
0   0     1       1  
1   0     2       1  
2   0     2       1

In [21]:
#length/size of the dataset
len(hd)

303

In [6]:
# get the feature (x) ready = all columns except target
x = hd.drop("target", axis=1)
x.head(3)

age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  slope  \
0   63    1   3       145   233    1        0      150      0      2.3      0   
1   37    1   2       130   250    0        1      187      0      3.5      0   
2   41    0   1       130   204    0        0      172      0      1.4      2   

   ca  thal  
0   0     1  
1   0     2  
2   0     2

In [8]:
# get the labels (y) ready = target column
y = hd.target
y.head(3)

0    1
1    1
2    1
Name: target, dtype: int64

# Always Remember! <br>
## NEVER  EVALUATE  or  TEST your models on DATA that it is LEARNT FROM -- that's why we split it into training and test sets.
(because if you do, it is like cheating in an examination)  <br>
![](https://media.giphy.com/media/A4CLbWb9o5rj2/giphy.gif)

### Splitting the data into test and training sets
Imagine it this way
* Test data = Final exam
* Training data = Mock exam (So we use the train_test_split first!)

In [15]:
from sklearn.model_selection import train_test_split
x_train, x_test, y_train, y_test = train_test_split(x,y, test_size=0.2) 
# test_size=0.2 means that we want our test dataset to be 20% of the overall data

In [16]:
#check the shapes of the new matrices just created. 
x_train.shape, x_test.shape, y_train.shape, y_test.shape

((242, 13), (61, 13), (242,), (61,))

Just notice,  <br>
242 = 80% of 303 <br>
61 = 20% of 303 <br>
So, training set is 80% of overall data and test set is 20% of the overall data  <br>
(I have mentioned this above too!)

### Filling (also called "Imputing") or disregarding missing values